File: src/main/kotlin/io/titandata/titan/providers/generic/Delete.kt
/*
 * Copyright (c) 2019 by Delphix. All rights reserved.
 */

package io.titandata.titan.providers.generic

import io.titandata.client.apis.CommitsApi

class Delete (
    private val commitsApi: CommitsApi = CommitsApi()
) {
    fun deleteCommit(repository: String, commit: String) {
        commitsApi.deleteCommit(repository, commit)
        println("$commit deleted")
    }

    fun deleteTags(repository: String, commit: String, tags: List<String>) {
        val commit = commitsApi.getCommit(repository, commit)
        val commitTags = (commit.properties.get("tags") as Map<String, String>).toMutableMap()

        for (t in tags) {
            if (t.contains("=")) {
                val key = t.substringBefore("=")
                val value = t.substringAfter("=")
                if (commitTags.containsKey(key) && commitTags[key] == value) {
                    commitTags.remove(key)
                }
            } else {
                commitTags.remove(t)
            }
        }

        val metadata = commit.properties.toMutableMap()
        metadata["tags"] = commitTags

        commitsApi.updateCommit(repository, io.titandata.models.Commit(id=commit.id, properties=metadata))
    }
}

File: internal/app/providers/common/Delete.go
package common

import (
	"fmt"
	client "github.com/titan-data/titan-client-go"
	"strconv"
	"strings"
)

func DeleteCommit(repo string, commit string, port int)  {
	cfg.BasePath = "http://localhost:" + strconv.Itoa(port)

	commitsApi.DeleteCommit(ctx, repo, commit)
	fmt.Println(commit + " deleted")
}

func DeleteTags(repo string, commit string, tags []string, port int) {
	cfg.BasePath = "http://localhost:" + strconv.Itoa(port)

	c, _, _ := commitsApi.GetCommit(ctx, repo, commit)
	cTags := c.Properties["tags"].(map[string]string)
	for _, t := range tags {
		if strings.Contains(t, "=") {
			s := strings.Split(t, "=")
			k := s[0]
			v := s[1]

			val, ok := cTags[k]
			if ok && val == v {
				delete(cTags, k)
			}
		} else {
			delete(cTags, t)
		}
	}
	metadata := Metadata{}.Load(c.Properties)
	metadata.SetTags(cTags)
	cm := client.Commit{
		Id:         c.Id,
		Properties: metadata.ToMap(),
	}
	commitsApi.UpdateCommit(ctx, repo, c.Id, cm)
}